### DISCUSSION TITLE
How to stop KPP 2.23 from crashing when a species reach very low concentration

### DISCUSSION THREAD

**Discussion by Fu-Weitao** [KineticPreProcessor/KPP#114] [Q&A]
I was using BOXMOX (https://mbees.med.uni-augsburg.de/gitlab/mbees/boxmox), which is built upon KPP 2.2.3. When using MOZART-T1 chemical mechanism, the program sometimes aborted, showing "Forced exit from Rosenbrock due to the following error: --> Step size too small: T + 10*H = T or H < Roundoff". I read https://gchp.readthedocs.io/en/stable/geos-chem-shared-docs/supplemental-guides/error-guide.html, and noticed that the error might be related to species concentrations or reaction rates being unusually high or low. So I looked into the output, and found there is indeed one species that reach a ppb value of ~0.38e-167.

I tried to replace the codes (drv, int, src, util) related to KPP 2.33 with KPP 2.6, ./compile, make, make install, but encountered some errors when compiling "MOZART_T1" that I was unable to resolve, because I'm unfamiliar with Fortran.

I wonder if there is any snippet in the newest KPP version that deals with program failure when some chemical species reach very low/high values, so that I can apply it as a patch to BOXMOX-KPP 2.2.3.

Thank you!

**Comment by yantosca** [comment]
Thanks for writing @Fu-Weitao.  We would recommend that you update to the latest KPP 3.1.1 anyway, since there have been several new features and fixes introduced since KPP 2.2.3.  

I am not familiar with BOXMOX so I cannot provide any insight.  What you might try is to adjust the tolerances ATOL and RTOL for the Rosenbrock solver.  You should be able to do this on a per-species basis.

**Reply by Fu-Weitao** [reply]
Hi, Thank you for replying. I'm trying to update it to KPP 2.6 first because I suppose it should be easier compared to upgrading to 3.1.1. But when running "VAR => C(1:NVAR )" or "FIX => C(NVAR+1:NSPEC)" the programs returns "Error: Non-POINTER in pointer association context (pointer assignment) & Error: Pointer assignment target is neither TARGET nor POINTER". I notice these two point assignments are added to multiple f90 files and freed shortly after. In KPP 2.2.3, these two steps do not exist.
I only replaced files in four folders (drv, int, src, util), and I'm not sure where the "C" is defined. Do you have any suggestions on which file I should look into? Thank you for your help!
